File: C&C++/Leetcode/8_StringtoInteger.c
#include <stdio.h>

int myAtoi(char *str)
{

    int n = 0;
    int ans = 0;

    while (str[n] != '\0' && str[n] == ' ')
        n++;

    if (str[n] == '\0' || !((str[n] <= '9' && str[n] >= '0') || (str[n] == '+' || str[n] == '-')))
        return 0;

    int flag = 1;
    if (str[n] == '-')
    {
        flag = -1;
        n++;
    }
    else if (str[n] == '+')
        n++;

    while (str[n] <= '9' && str[n] >= '0')
    {
        int t = str[n] - 48;
        if (flag == 1 && (ans > 214748364 || (ans == 214748364 && t >= 7)))
            return 2147483647;
        else if (flag == -1 && (ans > 214748364 || (ans == 214748364 && t >= 8)))
            return -2147483648;
        else
            ans = ans * 10 + t;
        n++;
    }

    return ans*flag;
}

int main()
{
    char c[] = "-2147483647";
    printf("%d", myAtoi(c));
}

File: Java/Leetcode/07_String/8_StringtoInteger.java

class Solution {

    public static void main(String[] args) {

        System.out.println(myAtoi("-91283472332"));

    }

    public static int myAtoi(String str) {

        str = str.trim();
        int len = str.length();
        if (len < 1)
            return 0;

        int i = 0;
        int flag = 1;
        int ans = 0;

        if (str.charAt(i) == '+')
            i++;
        else if (str.charAt(i) == '-') {
            i++;
            flag = -1;
        }

        while (i < len && str.charAt(i) <= '9' && str.charAt(i) >= '0') {
            if (flag > 0 && ((ans > Integer.MAX_VALUE / 10)
                    || ((ans == Integer.MAX_VALUE / 10) && (str.charAt(i) - '0') > 7)))
                return Integer.MAX_VALUE;
            if (flag < 0 && ((ans > Integer.MAX_VALUE / 10)
                    || ((ans == Integer.MAX_VALUE / 10) && (str.charAt(i) - '0') > 8)))
                return Integer.MIN_VALUE;

            ans = ans * 10 + str.charAt(i) - '0';
            i++;
        }

        return flag * ans;

    }

}
